### PR TITLE
Add NFD nightly test runs

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
Signed-off-by: Carlos Eduardo Arango Gutierrez <carangog@redhat.com>

### CODE REVIEW

File: ci-operator/config/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master__nfd.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  image_stream_tag:
+    name: release
+    namespace: openshift
+    tag: golang-1.16
+releases:
+  latest:
+    candidate:
+      product: ocp
+      stream: nightly
+      version: "4.8"
```
Comment (dagrayvid) [inline ci-operator/config/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master__nfd.yaml:16]:
```suggestion
      version: "4.8"
images:
- dockerfile_path: build/Dockerfile
  from: os
  to: ci-artifacts
```

**Comment by ArangoGutierrez** [comment]
Would love your eyes on this PR @kpouget

File: ci-operator/jobs/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master-periodics.yaml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  - base_ref: master
+    org: openshift-psap
+    repo: ci-artifacts
+  labels:
+    ci-operator.openshift.io/prowgen-controlled: "true"
+    ci-operator.openshift.io/variant: nfd
+    job-release: "4.8"
+    pj-rehearse.openshift.io/can-be-rehearsed: "true"
+  name: periodic-ci-openshift-psap-ci-artifacts-master-nfd-e2e-master
+  spec:
```
Comment (kpouget) [inline ci-operator/jobs/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master-periodics.yaml:18]: slack reporting missing here

File: ci-operator/config/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master__nfd.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+resources:
+  '*':
+    limits:
+      memory: 2Gi
+    requests:
+      cpu: 2000m
+      memory: 2Gi
+tests:
+- as: e2e-master
+  cron: 0 */23 * * *
```
Comment (kpouget) [inline ci-operator/config/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master__nfd.yaml:30]: in the general organization of the files / tests, I chose to:
1. not to run any periodic test in master`, only `presubmit` (so you can remove this `cron` stanza to follow the same pattern)
2. in `release-4.x`, run the nightly tests dedicated to OCP `4.x`
so this `4.8` testing should be moved to `release4.8__nfd`

File: ci-operator/jobs/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master-periodics.yaml (hunk 2 above)
Comment (kpouget) [reply ci-operator/jobs/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master-periodics.yaml:18]: (but the test will disappear from here if you follow the suggestion to remove periodics from `master`)

File: ci-operator/jobs/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master-presubmits.yaml [hunk 4]
```diff
@@ -451,6 +451,57 @@ presubmits:
         secret:
           secretName: result-aggregator
     trigger: (?m)^/test( | .* )images,?($|\s.*)
+  - agent: kubernetes
+    always_run: true
```
Comment (kpouget) [inline ci-operator/jobs/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master-presubmits.yaml:455]: no need to always run `nfd-image` (field is automatically set to `always_run: true`, we need to double check it each time new tests are added :))

File: ci-operator/config/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master__nfd.yaml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      version: "4.8"
+resources:
+  '*':
+    limits:
+      memory: 2Gi
+    requests:
+      cpu: 2000m
+      memory: 2Gi
+tests:
+- as: e2e-master
```
Comment (kpouget) [inline ci-operator/config/openshift-psap/ci-artifacts/openshift-psap-ci-artifacts-master__nfd.yaml:29]: could you name it `nfd-operator-e2e-master`, for consistency? (just a suggestion)

**Review by kpouget — CHANGES_REQUESTED** [review]
Hello @ArangoGutierrez, thanks for the PR, I'm looking forward to get NFD nightly tested with a ci-dashboard to report it :)

I understand that you only want to test on OCP 4.8.
I would suggest moving the test from `master__nfd.yaml` to `release-4.8__nfd.yaml` (see comment)
Don't forget to add slack reporting once the job file has been generated

I would also suggest having a `/test nfd-e2e-master` presubmit test, this time in `master__nfd.yaml`
and don't forget to mark this test as `always_run: false` in the generated job file

looking at the rehearsal, I need no error linked to *this PR* (`openshift/release`)

thanks :)

**Comment by kpouget** [comment]
```
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> + make image IMAGE_TAG=nfd-operator-ci:latest
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> make: go: No such file or directory
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> docker build -t nfd-operator-ci:latest \
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> 	 \
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> 	 ./
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> STEP 1: FROM registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.16-openshift-4.9 AS builder
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> Trying to pull registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.16-openshift-4.9...
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> STEP 2: FROM registry.ci.openshift.org/ocp/4.9:base
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> time="2021-06-22T00:12:18Z" level=error msg="unable to write build event: \"write unixgram @0001a->/run/systemd/journal/socket: sendmsg: no such file or directory\""
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> Error: error creating build container: Error initializing source docker://registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.16-openshift-4.9: Error reading manifest rhel-8-golang-1.16-openshift-4.9 in registry.ci.openshift.org/ocp/builder: unauthorized: authentication required
```
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/19246/rehearse-19246-periodic-ci-openshift-psap-ci-artifacts-master-nfd-e2e-master/1407117591662563328/artifacts/e2e-master/nightly/artifacts/002__nfd-operator__deploy_from_commit/_ansible.log

this one may not be relevant:
```
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> make: go: No such file or directory
```

this one seems to be an authentication missing?
```
2021-06-22 00:12:35,536 p=470 u=psap-ci-runner n=ansible | <stdout> Error: error creating build container: Error initializing source docker://registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.16-openshift-4.9: Error reading manifest rhel-8-golang-1.16-openshift-4.9 in registry.ci.openshift.org/ocp/builder: unauthorized: authentication required
```

```

2021-06-22 00:12:35,535 p=470 u=psap-ci-runner n=ansible | <stdout> + sed -i 's/registry.ci.openshift.org\/ocp\/builder:rhel-8-golang-1.16-openshift-4.8/docker.io\/library\/golang:1.16.3-buster/' Dockerfile
2021-06-22 00:12:35,535 p=470 u=psap-ci-runner n=ansible | <stdout> + sed -i 's/registry.ci.openshift.org\/ocp\/4.8:base/registry.access.redhat.com\/ubi8\/ubi/' Dockerfile
```
suggestion for improving this line, if you control where it comes from: use `se 's|<old str>|<new str>'` to avoid escaping the delimiter


I wonder why the testing working on my cluster without credentials to `registry.ci.openshift.org` ...?

**Comment by ArangoGutierrez** [comment]
This PR https://github.com/openshift-psap/ci-artifacts/pull/200 should fix rehearse

**Comment by kpouget** [comment]
thanks @ArangoGutierrez,
/lgtm

we'll use the CI to validate this one https://github.com/openshift-psap/ci-artifacts/pull/200

**Pull request merged**